PR: https://github.com/odoo/odoo/pull/233963

From: 2ad7ae8c10fba650687a084d823c695ee72271fe
From: Aurélien Warnon
Date: 2026-02-14 01:01:28

Breaking data model changes scores: del:2 + add:0, change matches:
-    name = fields.Char('Name', related='source_id.name', readonly=False)
-    source_id = fields.Many2one('utm.source', string='Source', required=True, ondelete='restrict', copy=False)

Total Changes: 65

[CLN] utm,test_mass_mailing: remove utm.source.mixin

RATIONALE

The way we currently use the UTM trackers (aka source/medium/campaign) does not
comply with the industry standards, leading to confusion and globally unclear
or not usable reporting on UTM data.

PURPOSE

Now that all functional modules have been modified to remove the usage of
utm.source.mixin, we can get rid of it once and for all.

See other commits for case by case details.

Task-3330036

Part-of: odoo/odoo#233963
Related: odoo/enterprise#98569
Related: odoo/upgrade#8746
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/utm/models/utm_source.py
+++ b/addons/utm/models/utm_source.py
@@ -57,68 +57,3 @@ class UtmSource(models.Model):
                     "Doing so would be like tearing down a load-bearing wall \u2014 not the best idea.",
                     utm_source.name
                 ))
-
-
-class UtmSourceMixin(models.AbstractModel):
-    """Mixin responsible of generating the name of the source based on the content
-    (field defined by _rec_name) of the record (mailing, social post,...).
-    """
-    _name = 'utm.source.mixin'
-    _description = 'UTM Source Mixin'
-
-    name = fields.Char('Name', related='source_id.name', readonly=False)
-    source_id = fields.Many2one('utm.source', string='Source', required=True, ondelete='restrict', copy=False)
-
-    @api.model
-    def default_get(self, fields):
-        # Exclude 'name' from fields to avoid retrieving it from context.
-        return super().default_get([field for field in fields if field != "name"])
-
-    @api.model_create_multi
-    def create(self, vals_list):
-        """Create the UTM sources if necessary, generate the name based on the content in batch."""
-        # Create all required <utm.source>
-        utm_sources = self.env['utm.source'].create([
-            {
-                'name': values.get('name')
-                or self.env.context.get('default_name')
-                or self.env['utm.source']._generate_name(self, values.get(self._rec_name)),
-            }
-            for values in vals_list
-            if not values.get('source_id')
-        ])
-
-        # Update "vals_list" to add the ID of the newly created source
-        vals_list_missing_source = [values for values in vals_list if not values.get('source_id')]
-        for values, source in zip(vals_list_missing_source, utm_sources):
-            values['source_id'] = source.id
-
-        for values in vals_list:
-            if 'name' in values:
-                del values['name']
-
-        return super().create(vals_list)
-
-    def write(self, vals):
-        if (vals.get(self._rec_name) or vals.get('name')) and len(self) > 1:
-            raise ValueError(
-                _('You cannot update multiple records with the same name. The name should be unique!')
-            )
-
-        if vals.get(self._rec_name) and not vals.get('name'):
-            vals['name'] = self.env['utm.source']._generate_name(self, vals[self._rec_name])
-        if vals.get('name'):
-            vals['name'] = self.env['utm.mixin'].with_context(
-                utm_check_skip_record_ids=self.source_id.ids
-            )._get_unique_names("utm.source", [vals['name']])[0]
-
-        return super().write(vals)
-
-    def copy_data(self, default=None):
-        """Increment the counter when duplicating the source."""
-        default = default or {}
-        default_name = default.get('name')
-        vals_list = super().copy_data(default=default)
-        for source, vals in zip(self, vals_list):
-            vals['name'] = self.env['utm.mixin']._get_unique_names("utm.source", [default_name or source.name])[0]
-        return vals_list
